### DISCUSSION TITLE
Bug? Strange spot in turbulence field on the wall at the edge of the mpi partition

### DISCUSSION THREAD

**Discussion by bigfooted** [su2code/SU2#1768] [General]
I did some more simulations on my 90 degree pipe bend and I noticed some spots in y+, tke and omega. This is a test for wall functions by the way, where I impose the values for tke and omega. 
A picture is here:
https://github.com/bigfooted/su2cases/blob/master/validation/sudo_pipe_bend/paraview_screenshot_hotspot_tke.png

It looks like these spots appear at the edge of the mpi partitions. A further examination showed that these high tke values only appear on halo nodes, where we have:
```(!geometry->nodes->GetDomain(iPoint))```

Not sure how to get rid of these spots. Since they are on halo nodes, do they then not affect the solution and it is only a visualization thing?

**Comment by pcarruscag** [comment]
We talked about this in the context of a fix by @maxaehle (illegal access to normal neighbor IIRC)
Do you get exactly the same solution in serial and parallel?

**Reply by bigfooted** [reply]
Finally had some more time to look at this... The issue was mentioned in #1430. 
It looks like it is related to this issue. It also looks like it is more than a visualization issue, since tke downstream is affected by the spots in the tke at the edges of the mpi partitions. The origin of the problem is YPlus, UTau and EddyViscWall, so the computations in CIncNSSolver::SetTauWall_WF

My understanding of do's and don't wrt halo nodes is limited. In CTurbSSTSolver::SetTurbVars_WF for instance, we do not skip halo nodes, which we do skip in the computation of the Y+ in CIncNSSolver::SetTau_Wall_WF. This at least seems inconsistent. So we have to skip halo nodes in SetTurbVars_WF? But don't we need to set something at the halo nodes?

**Reply by bigfooted** [reply]
So, if I compute y+ etc at halo nodes in SetTauWall_WF ( I commented the line` if (!geometry->nodes->GetDomain(iPoint)) continue;` ) then there are no spots anymore in the solution.
But this would then be a problem for full MG problems?

**Reply by pcarruscag** [reply]
If the information is all there to compute on halo nodes then that is the best option.
You can skip halos only on coarse levels, since it should not matter for the final solution, or for robustness it may even be better to not use wall functions on coarse levels, just setting safe values.

**Reply by bigfooted** [reply]
OK, I will then compute on halo nodes as well. Why would it be not robust to use wall functions on coarse levels? With turbulence, I'd say wall functions would be good on coarse levels as well since the y+ is larger on the coarse levels. 

Is this (potential issues with too coarse mesh near the wall) not the reason that turbulence uses SingleGrid_Iteration?

**Reply by pcarruscag** [reply]
You can check MG_Level of the solver for MESH_0, no need to pass iMesh around.
Anything that requires gradients is a robustness problem on coarse levels because the mesh is bad.
No, I think the reason is simplicity.

**Reply by dylanrubini** [reply]
Hi @pcarruscag @bigfooted ! I was wondering whether there was any update on this discussion? I am encountering the same issue  for a case running SA model with wall functions on the Onera M6 wing. The spikes are seen in the eddy viscosity (but same issue occurs with SST model) and occur with/without multi-grid. The problem only happens directly at the surface and only occurs with wall functions. I am running on latest master branch. Let me know any potential solutions please? Note that the impact of spikes in eddy viscosity does propagate to other quantities like temperature on the surface (leading to temp spikes). Thanks!

**Reply by bigfooted** [reply]
This issue was fixed in #1791 Can you share your mesh and config file? Then I can reproduce the issue and see if this is related.

**Reply by dylanrubini** [reply]
@bigfooted thanks for getting back to me quickly. For reference I have posted a figure below. I have also attached the configuration file _(note the extension change to get GitHub to accept the file)_ 

Is there somewhere I could upload the mesh file maybe ? it is quite big

<img width="720" height="693" alt="image" src="https://github.com/user-attachments/assets/d04614da-d840-43c3-a435-aea7b9af2be7" />

[wing_su2_case.txt](https://github.com/user-attachments/files/23121005/wing_su2_case.txt)

**Reply by dylanrubini** [reply]
@bigfooted I made a smaller mesh file and uploaded it here https://github.com/dylanrubini/spikes_su2.git 

Side note: I run on 72 cores if that is relevant since this issue may be related to halo nodes.

**Reply by dylanrubini** [reply]
@bigfooted At first glance, it appears that the issue persists on single core

**Reply by dylanrubini** [reply]
@bigfooted further to this, I can also reproduce the same spikes in the eddy viscosity when using the SU2 Onera M6 Turbulent Compressible flow tutorial and simply adding wall functions - https://su2code.github.io/tutorials/Turbulent_ONERAM6/

**Reply by bigfooted** [reply]
the Onera tutorial shows these spots because the Newton method for y+ did not converge. At these spots, y+ = 0. Maybe because the mesh is really coarse? In any case, y+=0 should not happen. 
I could make the spots disappear by using much more iterations for the newton solver:
```
MARKER_WALL_FUNCTIONS=(WING, STANDARD_WALL_FUNCTION)
WALLMODEL_MINYPLUS=0.5
WALLMODEL_MAXITER=1000
WALLMODEL_RELFAC= 0.1
```
For your 3D testcase, is the wallfunction model complaining that it did not converge?

**Reply by dylanrubini** [reply]
@bigfooted thanks for taking a look. I also took the same strategy as you, and still  got the same results with same spikes for the wing I was running

**Reply by dylanrubini** [reply]
I do get these warnings "Warning: y+ < 0.5 in 1 points, for which the wall model is not active." sometimes. but ultimately the minimum yplus in the domain ends up being 3.9

**Reply by bigfooted** [reply]
Fix: #2599

**Reply by dylanrubini** [reply]
@bigfooted I can confirm this also addresses the issue for me. thanks for the prompt fix!!

**Reply by bigfooted** [reply]
Great! Thanks for helping us fix this!
